scripts/diff2patches.1
.TH "diff2patches" "1" "" "Raphael Geissert <atomo64@gmail.com>" ""
.SH "NAME"
.LP
diff2patches \- Extract non\-debian/ patches from .diff.gz files
.SH "SYNTAX"
.LP
\fBdiff2patches \fIfilename\fP
.br
\fBdiff2patches \-\-help\fR|\fB\-\-version\fP
.SH "DESCRIPTION"
.LP
Extracts patches from .diff.gz which apply to files outside the
\*(lqdebian/\*(rq directory scope.  A patch is created for each modified file.
Each patch is named according to the path of the modified file, with \*(lq/\*(rq
replaced by \*(lq___\*(rq, and an extension of \*(lq.patch\*(rq.
.SH "OPTIONS"
.LP
.TP 4
\fB\fIfilename\fP\fR
Extract patches from \fB\fIfilename\fP\fR which apply outside the
\*(lqdebian/\*(rq directory.
.TP
\fB\-\-help\fR
Output help information and exit.
.TP
\fB\-\-version\fR
Output version information and exit.
.SH "FILES"
.TP
\fIdebian/control\fP
Existence of this file is tested before any patch is extracted.
.TP
\fIdebian/\fP
.TQ
\fIdebian/patches/\fP
Patches are extracted to one of these directories.  \*(lqdebian/patches/\*(rq is
preferred, if it exists.  If \fIDEB_PATCHES\fP is present in the environment,
it will override this behavior (see \*(lqENVIRONMENT VARIABLES\*(rq section
below).
.SH "ENVIRONMENT VARIABLES"
.TP
\fBDEB_PATCHES\fP
When defined and points to an existing directory, patches are extracted in that directory
and not under \*(lqdebian/\*(rq nor \*(lqdebian/patches/\*(rq.
.SH "SEE ALSO"
.LP
\fBcombinediff\fR(1)
.SH "AUTHOR"
.LP
Raphael Geissert <atomo64@gmail.com>


debian/NEWS
devscripts (2.10.54) unstable; urgency=low

    manpage-alert now, as originally intended, prints out the full path to the
    binary which is lacking a manpage instead of simply the binary name.

    Before:
      No manual entry for somebinary

    After:
      No manual entry for /usr/bin/somebinary

 -- James Vega <jamessan@debian.org>  Thu, 06 Aug 2009 18:03:41 -0400

devscripts (2.10.24) unstable; urgency=low

    Starting with devscripts 2.10.21, debchange attempts to
    automatically determine whether an NMU changelog should be
    started, by comparing DEBEMAIL and DEBFULLNAME to the
    package's Maintainer and Uploaders fields.

    This behaviour may now be disabled using the DEBCHANGE_AUTO_NMU
    configuration file variable.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Mon, 07 Apr 2008 14:38:08 +0100

devscripts (2.10.20) unstable; urgency=low

    devscripts 2.10.19 modified debcommit's --release option to use the
    uncommitted changes (if any) in the changelog to derive the commit
    message.

    This behaviour is now optional and can be enabled using the -R /
    --release-use-changelog command line option or by setting the
    DEBCOMMIT_RELEASE_USE_CHANGELOG configuration option to "yes".

    The default is to use the previous (i.e. pre-2.10.19) behaviour.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Wed, 19 Mar 2008 20:31:20 +0000

devscripts (2.10.19) unstable; urgency=low

    In devscripts 2.10.17, dget was modified to unpack source packages by
    default.  The new -d switch or the DGET_UNPACK configuration file option
    may be used to indicate that the unpacking should not occur.

    As linda has been removed from both unstable and testing, support for
    it has been removed from debuild and cvs-debuild.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Thu, 13 Mar 2008 19:33:38 +0000


scripts/diff2patches.sh
#!/bin/bash -e

####################
#    Copyright (C) 2007, 2008 by Raphael Geissert <atomo64@gmail.com>
#
#    This file is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This file is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this file  If not, see <http://www.gnu.org/licenses/>.
#
#    On Debian systems, the complete text of the GNU General
#    Public License 3 can be found in '/usr/share/common-licenses/GPL-3'.
####################

PROGNAME=$(basename "$0")

usage () {
    echo \
"Usage: $PROGNAME [options] FILE.diff.gz
  Options:
    --help          Show this message
    --version       Show version and copyright information
  debian/control must exist on the current path for this script to work
  If debian/patches exists and is a directory, patches are extracted there,
  otherwise they are extracted under debian/ (unless the environment variable
  DEB_PATCHES is defined and points to a valid directory, in which case
  patches are extracted there)."
}

version () {
    echo \
"This is $PROGNAME, from the Debian devscripts package, version ###VERSION###
This code is copyright 2007, 2008 by Raphael Geissert, all rights reserved.
This program comes with ABSOLUTELY NO WARRANTY.
You are free to redistribute this code under the terms of the
GNU General Public License, version 3 or later."
}

case "$1" in
	--help) usage; exit 0 ;;
	--version) version; exit 0 ;;
esac

if ! which lsdiff >/dev/null 2>&1; then
	echo "lsdiff was not found in \$PATH, package patchutils probably not installed!"
	exit 1
fi

diffgz="$1"

if [ ! -f "$diffgz" ]; then
	[ -z "$diffgz" ] && diffgz="an unspecified .diff.gz"
	echo "Couldn't find $diffgz, aborting!"
	exit 1
fi

if [ -x /usr/bin/dh_testdir ]; then
	/usr/bin/dh_testdir || exit 1
else
	[ ! -f debian/control ] && echo "Couldn't find debian/control!" && exit 1
fi

if [ -z "$DEB_PATCHES" ] || [ ! -d "$DEB_PATCHES" ]; then
	DEB_PATCHES=debian
	[ -d debian/patches ] && DEB_PATCHES=debian/patches
else
	DEB_PATCHES="$(readlink -f "$DEB_PATCHES")"
fi

echo "Patches will be extracted under $DEB_PATCHES/"

FILES=$(zcat "$diffgz" | lsdiff --strip 1 | egrep -v ^debian/) || \
	echo "$(basename "$diffgz") doesn't contain any patch outside debian/"

for file in $FILES; do
	[ ! -z "$file" ] || continue
	echo -n "Extracting $file..."
	newFileName="$DEB_PATCHES/$(echo "$file" | sed 's#/#___#g').patch"
	zcat "$diffgz" | filterdiff -i "$file" -p1 > "$newFileName"
	echo "done"
done

exit
